### PR TITLE
[NET-6593] agent: check for minimum RSA key size

### PR DESCRIPTION
**Pull request by loshz**
### Description

When Auto-Config is enabled for a client agent, it uses the `generateCSR` function to create a certificate signing request (CSR) self-signed by a locally generated RSA or ECDSA private key.

To create the key, `generateCSR` reads the `private_key_type` and `private_key_bits` options from the client’s Connect configuration. If `private_key_type` is not configured, `generateCSR` defaults to creating an ECDSA key. If `private_key_bits` is not configured, generateCSR sets the key length to 256 bits, which is an appropriate size for an ECDSA key.

A 256-bit RSA key is very weak and can be cracked easily using an integer factorization library such as Msieve. The impact of this issue is limited because, although the weak key is generated, its small size prevents signing and encrypting messages longer than 256 bits, which causes the agent to exit with an error during creation of the CSR.

This PR updates the `generateCSR`  to check for the key type, and sets a key size accordingly.

### Links

- https://hashicorp.atlassian.net/browse/NET-6593

### PR Checklist

* [x] updated test coverage
* [ ] external facing docs updated
* [x] appropriate backport labels added
* [ ] not a security concern

### CODE REVIEW

File: agent/auto-config/tls.go [hunk 1]
```diff
@@ -241,7 +241,12 @@ func (ac *AutoConfig) generateCSR() (csr string, key string, err error) {
 		conf.PrivateKeyType = connect.DefaultPrivateKeyType
 	}
 	if conf.PrivateKeyBits == 0 {
-		conf.PrivateKeyBits = connect.DefaultPrivateKeyBits
+		// If using an RSA key, a key size of at least 2048 bits is recommended; 4096 bits is better.
+		if conf.PrivateKeyType == connect.PrivateKeyTypeRSA {
+			conf.PrivateKeyBits = connect.DefaultPrivateKeyBitsRSA
+		} else {
+			conf.PrivateKeyBits = connect.DefaultPrivateKeyBits
+		}
```
Comment (zalimeni) [inline agent/auto-config/tls.go:249]: ~ One other option here might be to allow passing `0` through to `GeneratePrivateKeyWithConfig` and let it negotiate the default based on the type. That might avoid repeating this logic other places or missing a 3rd key type -> bits default in the future. Happy to wait on that if you think it's best to keep the defaulting here and explicit for now.

**Review by zalimeni — APPROVED** [review]

File: agent/connect/generate.go [hunk 2]
```diff
@@ -21,6 +21,11 @@ const (
 	DefaultPrivateKeyType      = "ec"
 	DefaultPrivateKeyBits      = 256
 	DefaultIntermediateCertTTL = 24 * 365 * time.Hour
+
+	// RSA specific settings.
+	PrivateKeyTypeRSA        = "rsa"
```
[the remark this answers is not in the document]
Comment (loshz) [reply agent/connect/generate.go:26]: Do you mean updating [these](https://github.com/hashicorp/consul/blob/main/agent/connect/generate.go#L90-L92) values to use the above constants?

File: agent/auto-config/tls.go (hunk 1 above)
Comment (loshz) [reply agent/auto-config/tls.go:249]: I like the idea of using the defaults, but I also like being specific, too. Happy to do either.

File: agent/connect/generate.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (zalimeni) [reply agent/connect/generate.go:26]: Exactly!

File: agent/auto-config/tls.go (hunk 1 above)
Comment (zalimeni) [reply agent/auto-config/tls.go:249]: No strong preference, glad to go w/ what you have here - more thinking out loud and noticing we haven't really centralized that type-defaulting logic. But the validation being centralized (which you have done) is the important bit.

**Review by dhiaayachi — APPROVED** [review]
Thank you for working on this @loshz!
🚀

File: agent/connect/generate.go [hunk 3]
```diff
@@ -35,6 +40,11 @@ func pemEncode(value []byte, blockType string) (string, error) {
 func generateRSAKey(keyBits int) (crypto.Signer, string, error) {
 	var pk *rsa.PrivateKey
 
+	// Check for a secure key length.
+	if keyBits < MinPrivateKeyBitsRSA {
+		return nil, "", fmt.Errorf("error generating RSA private key: key size must be at least %d bits", MinPrivateKeyBitsRSA)
```
Comment (analogue) [inline agent/connect/generate.go:45]: UX: would be nice to surface the `"got: %d", keyBits` as part of the error message

**Pull request merged**